python/mpt/src/mpt/__init__.py

```diff
@@ -0,0 +1,1372 @@
+from ethereum.cancun.trie import (
+    BranchNode,
+    ExtensionNode,
+    InternalNode,
+    LeafNode,
+    Trie,
+    bytes_to_nibble_list,
+    nibble_list_to_compact,
+    trie_set,
+)
+from ethereum.crypto.hash import Hash32, keccak256
+from ethereum.utils.hexadecimal import hex_to_bytes
+from ethereum_rlp import rlp
+from ethereum_types.bytes import Bytes, Bytes32
+from ethereum_types.numeric import U256, Uint
+
+from mpt.state_diff import StateDiff
+
+logger = logging.getLogger("mpt")
+logger.propagate = False
+logger.setLevel(logging.DEBUG)
+if not logger.handlers:
+    ch = logging.StreamHandler()
+    ch.setLevel(logging.DEBUG)
+    formatter = logging.Formatter(
+        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+    )
+    ch.setFormatter(formatter)
+    logger.addHandler(ch)
+
+
+EMPTY_TRIE_ROOT_HASH = Hash32(
+    bytes.fromhex("56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
+)
+
+EMPTY_CODE_HASH = Hash32(
+    bytes.fromhex("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470")
+)
+
+EMPTY_BYTES_RLP = b"\x80"
+
+
+def nibble_path_to_hex(nibble_path: Bytes) -> Bytes:
+    return (
+        "0x"
+        + Bytes(
+            bytes(
+                [
+                    nibble_path[i] * 16 + nibble_path[i + 1]
+                    for i in range(0, len(nibble_path) - 1, 2)
+                ]
+            )
+        ).hex()
+    )
+
+
+class EthereumState:
+    """
+    Represents the (partial) state of Ethereum for a given block.
+    Includes all account MPT leaf nodes and storage MPT leaf nodes and all intermediary nodes touched in a block,
+    related codes, an access list of addresses and storage keys that were accessed in the block, and a state root.
+    """
+
+    nodes: Mapping[Bytes32, Bytes]
+    codes: Mapping[Bytes32, Bytes]
+    access_list: Mapping[Address, Optional[List[Bytes32]]]
+    state_root: Hash32
+
+    @classmethod
+    def create_empty(cls) -> "EthereumState":
+        return cls(
+            nodes={},
+            codes={},
+            access_list={},
+            state_root=EMPTY_TRIE_ROOT_HASH,
+        )
+
+    def __init__(
+        self,
+        nodes: Mapping[Bytes32, Bytes],
+        codes: Mapping[Bytes32, Bytes],
+        access_list: Mapping[Address, Optional[List[Bytes32]]],
+        state_root: Hash32,
+    ):
+        self.nodes = nodes
+        self.codes = codes
+        self.access_list = access_list
+        self.state_root = state_root
+        logger.debug(f"Initialized MPT with state root: 0x{state_root.hex()}")
+        logger.debug(f"Number of nodes: {len(nodes)}")
+        logger.debug(f"Number of codes: {len(codes)}")
+        logger.debug(f"Number of addresses in access list: {len(access_list)}")
+
+    def to_state(self) -> State:
+        """
+        Convert the EthereumState to a State object.
+
+        Returns
+        -------
+        State
+            A State object representing the Ethereum state
+        """
```

Comment:
see here -> State object representing Ethereum State, on a class named EthereumState, that returns State 🥴